File: southampton_crime_map_backend/app/services/routing_service.py
"""OpenRouteService routing client."""

import asyncio
import hashlib
import json
import logging
from typing import Any, Dict, List, Optional

import httpx
import redis.asyncio as redis

from app.config import get_settings
from app.core.exceptions import ExternalServiceError

logger = logging.getLogger(__name__)
settings = get_settings()


class RoutingService:
    """OpenRouteService routing client with Redis caching."""

    def __init__(self):
        self.base_url = settings.ORS_API_URL
        self.api_key = settings.ORS_API_KEY
        self.timeout = 15.0
        self.max_retries = 2
        self.cache_ttl = 86400  # 24 hours
        self._redis_client: Optional[redis.Redis] = None

    async def _get_redis_client(self) -> Optional[redis.Redis]:
        """Get or create Redis client."""
        if self._redis_client is None:
            try:
                self._redis_client = redis.from_url(
                    settings.REDIS_URL, encoding="utf-8", decode_responses=True
                )
                # Test connection
                await self._redis_client.ping()
                logger.info("Redis connection established for ORS caching")
            except Exception as e:
                logger.warning(f"Redis connection failed: {str(e)}. Caching disabled.")
                self._redis_client = None
        return self._redis_client

    def _generate_cache_key(
        self, profile: str, coordinates: List[List[float]], alternatives: int
    ) -> str:
        """Generate cache key for ORS request."""
        data = f"{profile}:{alternatives}:{json.dumps(coordinates)}"
        return f"ors:{hashlib.md5(data.encode()).hexdigest()}"

    async def get_directions(
        self,
        coordinates: List[List[float]],  # [[lng, lat], [lng, lat]]
        profile: str = "foot-walking",
        alternatives: int = 3,
    ) -> Dict[str, Any]:
        """Get directions from OpenRouteService.

        Args:
            coordinates: [longitude, latitude] pairs
            profile: foot-walking, cycling-regular, or driving-car
            alternatives: Number of route alternatives

        Returns:
            GeoJSON response with route features

        Raises:
            ExternalServiceError: ORS unavailable or error
        """
        # Try to get from cache first
        cache_key = self._generate_cache_key(profile, coordinates, alternatives)
        redis_client = await self._get_redis_client()

        if redis_client:
            try:
                cached = await redis_client.get(cache_key)
                if cached:
                    logger.info(f"Cache HIT for {cache_key}")
                    return json.loads(cached)
                logger.info(f"Cache MISS for {cache_key}")
            except Exception as e:
                logger.warning(f"Redis get error: {str(e)}")

        # Fetch from ORS API
        url = f"{self.base_url}/v2/directions/{profile}/geojson"

        headers = {
            "Authorization": self.api_key,
            "Content-Type": "application/json",
        }

        body = {
            "coordinates": coordinates,
            "instructions": True,
            "alternative_routes": {
                "target_count": alternatives,
                "share_factor": 0.6,
                "weight_factor": 1.4,
            },
        }

        for attempt in range(self.max_retries):
            try:
                async with httpx.AsyncClient(timeout=self.timeout) as client:
                    response = await client.post(url, json=body, headers=headers)

                    if response.status_code == 200:
                        data = response.json()
                        logger.info(f"Fetched {len(data.get('features', []))} routes from ORS")

                        # Cache the response
                        if redis_client:
                            try:
                                await redis_client.setex(
                                    cache_key, self.cache_ttl, json.dumps(data)
                                )
                                logger.info(f"Cached ORS response for {cache_key}")
                            except Exception as e:
                                logger.warning(f"Redis set error: {str(e)}")

                        return data

                    elif response.status_code == 400:
                        logger.error(f"Invalid ORS request: {response.text}")
                        raise ExternalServiceError("Invalid routing request")

                    elif response.status_code == 429:
                        logger.warning("ORS rate limit exceeded")
                        if attempt < self.max_retries - 1:
                            await asyncio.sleep(2**attempt)
                            continue
                        raise ExternalServiceError("Rate limit exceeded")

                    else:
                        logger.error(f"ORS error {response.status_code}: {response.text}")
                        if attempt < self.max_retries - 1:
                            await asyncio.sleep(2**attempt)
                            continue
                        raise ExternalServiceError("Routing service unavailable")

            except httpx.TimeoutException:
                logger.error(f"ORS timeout (attempt {attempt + 1})")
                if attempt < self.max_retries - 1:
                    await asyncio.sleep(2**attempt)
                    continue
                raise ExternalServiceError("Routing service timeout")

            except Exception as e:
                logger.error(f"Error fetching routes: {str(e)}")
                raise ExternalServiceError(f"Routing error: {str(e)}")

        raise ExternalServiceError("Failed to fetch routes after retries")

    def extract_route_info(self, feature: Dict[str, Any]) -> Dict[str, Any]:
        """Extract route information from ORS GeoJSON feature.

        Args:
            feature: GeoJSON feature from ORS response

        Returns:
            Dict with route info (geometry, distance, duration, instructions)
        """
        properties = feature.get("properties", {})
        summary = properties.get("summary", {})

        return {
            "geometry": feature.get("geometry"),
            "distance_m": int(summary.get("distance", 0)),
            "duration_s": int(summary.get("duration", 0)),
            "instructions": properties.get("segments", [{}])[0].get("steps", []),
        }

File: nest-saferoute-backend/src/modules/route/services/routing.service.ts
import { Injectable, Logger, HttpException, HttpStatus } from '@nestjs/common';
import { ConfigService } from '@nestjs/config';
import { HttpService } from '@nestjs/axios';
import { firstValueFrom } from 'rxjs';
import * as turf from '@turf/turf';
import { LineString } from 'geojson';
import * as polyline from '@mapbox/polyline';

export interface ORSRoute {
  geometry: any;
  distance: number;
  duration: number;
  instructions?: any[];
}

@Injectable()
export class RoutingService {
  private readonly logger = new Logger(RoutingService.name);
  private readonly apiKey: string;
  private readonly apiUrl: string;

  constructor(
    private configService: ConfigService,
    private httpService: HttpService,
  ) {
    this.apiKey = this.configService.get<string>('externalApis.orsApiKey') || '';
    this.apiUrl = this.configService.get<string>('externalApis.orsApiUrl') || 'https://api.openrouteservice.org';
  }

  /**
   * Get route alternatives from OpenRouteService
   */
  async getRouteAlternatives(
    originLng: number,
    originLat: number,
    destLng: number,
    destLat: number,
    mode: string = 'foot-walking',
    alternatives: number = 3,
  ): Promise<ORSRoute[]> {
    try {
      const url = `${this.apiUrl}/v2/directions/${mode}`;

      const requestBody = {
        coordinates: [
          [originLng, originLat],
          [destLng, destLat],
        ],
        alternative_routes: {
          target_count: alternatives,
          weight_factor: 1.4,
          share_factor: 0.6,
        },
        instructions: true,
        elevation: false,
      };

      this.logger.debug(`Requesting routes from ORS: ${url}`);

      const response = await firstValueFrom(
        this.httpService.post(url, requestBody, {
          headers: {
            'Authorization': this.apiKey,
            'Content-Type': 'application/json',
          },
        }),
      );

      const routes = response.data.routes || [];

      this.logger.debug(`ORS returned ${routes.length} routes`);
      if (routes.length > 0) {
        this.logger.debug(`First route geometry type: ${typeof routes[0].geometry}, is string: ${typeof routes[0].geometry === 'string'}`);
      }

      return routes.map((route: any) => {
        // Decode polyline if geometry is a string
        let geometry = route.geometry;
        if (typeof geometry === 'string') {
          // Polyline returns [[lat, lng], ...], but GeoJSON needs [[lng, lat], ...]
          const decoded = polyline.decode(geometry);
          geometry = {
            type: 'LineString',
            coordinates: decoded.map((coord: [number, number]) => [coord[1], coord[0]]), // Swap lat/lng to lng/lat
          };
        }

        return {
          geometry,
          distance: route.summary?.distance || 0,
          duration: route.summary?.duration || 0,
          instructions: route.segments?.[0]?.steps || [],
        };
      });
    } catch (error) {
      this.logger.error(`ORS API error: ${error.message}`);
      if (error.response) {
        this.logger.error(`ORS response status: ${error.response.status}`);
        this.logger.error(`ORS response data: ${JSON.stringify(error.response.data).substring(0, 500)}`);
      }

      // Fallback: create a simple direct route
      this.logger.warn('Falling back to direct route');
      return [this.createDirectRoute(originLng, originLat, destLng, destLat)];
    }
  }

  /**
   * Create a simple direct route as fallback
   */
  private createDirectRoute(
    originLng: number,
    originLat: number,
    destLng: number,
    destLat: number,
  ): ORSRoute {
    const line = turf.lineString([
      [originLng, originLat],
      [destLng, destLat],
    ]);

    const distance = turf.length(line, { units: 'meters' });

    // Estimate walking duration at 5 km/h
    const duration = (distance / 1000) * (3600 / 5);

    return {
      geometry: line.geometry,
      distance,
      duration,
      instructions: [],
    };
  }

  /**
   * Validate coordinates
   */
  validateCoordinates(lat: number, lng: number): boolean {
    return lat >= -90 && lat <= 90 && lng >= -180 && lng <= 180;
  }

  /**
   * Calculate distance between two points
   */
  calculateDistance(
    lat1: number,
    lng1: number,
    lat2: number,
    lng2: number,
  ): number {
    const from = turf.point([lng1, lat1]);
    const to = turf.point([lng2, lat2]);
    return turf.distance(from, to, { units: 'meters' });
  }
}
